Onion routing chapter release prep

aantonop opened the pull request.

aantonop committed:
Intro onion routing

aantonop committed:
architecture highlight

aantonop committed:
rename and README update

aantonop committed:
onion routing path and payloads

aantonop committed:
onion payloads, key gen, ECDH

aantonop committed:
wrapping the onion

aantonop committed:
Wrapping the onion done

aantonop committed:
Breakdown wrapping the onion into more steps

aantonop committed:
misc edits and heading changes, README update

aantonop committed:
Unwrapping the onion

aantonop committed:
Unwrapping the onion diagrams

aantonop committed:
Error return and conclusion

aantonop committed:
Failure table sample

aantonop committed:
Small edit

aantonop committed:
Updated failure_types table

aantonop committed:
Python script to update failure type table from Github BOLT#4 spec

aantonop committed:
indent

aantonop committed:
remove unwanted file

Roasbeef committed:
ch-10: add errata about htlc min+max also being included in chan updates

Roasbeef committed:
ch-10: minor grammar fix for Alice payload construction

Roasbeef committed:
ch-10: mention the role of the payment_secret in e2e security

Without this value, any intermediate node can construct a _new_ onion
packet with modified values for the CLTV and amount, to attempt to see
if the next hop is actually the final hop in the route.

Roasbeef committed:
ch-10: typo fix for total_msat description

Roasbeef committed:
ch-10: explcitily call out all zero for scid as final hop indicator

In practice, the HMAC for the _next_ hop is also zero, but most
implementations will check the `scid` value instead.

Roasbeef commented:
I found it easier to do edits/review by branching off your current version and adding commits on top of that: https://github.com/aantonop/lnbook/compare/onion_routing_release...Roasbeef:onion-chapter-edits?expand=1 

Shooting to finish getting through the entire thing tomorrow, but from here we can either have me make a PR to your personal branch, or cherry-pick my commits in from there.

Roasbeef committed:
ch-10: small typo fix

Roasbeef committed:
ch-10: elaborate more on the usage of the rho key in the scheme

Roasbeef committed:
ch-10: mention session key re-blinding scheme at the core of sphinx

Roasbeef committed:
ch-10: mention that the inner HMAC for the final hop is all zeros

Roasbeef committed:
ch-10: specify that mu key is used to generate the inner HMAC

Roasbeef committed:
ch-10: add extra description of the 2 ways to recognize a terminal packet

Roasbeef committed:
ch-10: explain the distinction between the outer and inner HMACs

Roasbeef committed:
ch-10: fix ordering of HMAC steps, HMAC is computed over encrypted payload

The outer HMAC for the final hop is computed over the encrypted payload. Instead, the inner HMAC is just blank for the final hop. In this commit, we switch the ordering to reflect this. The diagrams will need to be updated as well.

Roasbeef committed:
ch-10: add section on replay protection+detection

Roasbeef committed:
ch-10: propagate ordering fix to the other hops

Roasbeef committed:
ch-10: bob only verifies the HMAC once, but he extracts the outer HAMC for chan

Continuing to propagate HMAC fix, Bob does't need to verify the HMAC again, instead he needs to extract the inner HMAC, as it'll become the outer HMAC for Chan.

Roasbeef committed:
ch-10: expand high level explanation of how Bob constructs the next onion packet

Still need to expand on the blinded of the session key as noted in the TODO section below.

Roasbeef committed:
ch-10: chan extracts the inner HMAC instead of verifying again

Roasbeef commented:
Alright, I've finished an initial pass all the way through. Excellent work on this chapter @aantonop! Really wish I had such a cogent explanation of the Sphinx packet construction+processing back in 2015 when I spent many nights/weekends at the desk in my dorm room trying to figure out how tf things actually worked 😅. The diagrams are amazing IMO, and really help to guide the user through the entire process. By the time they get to the later hops in the route it really just becomes "rinse and repeat". 

During editing I added a section on replay protection+detection since that wasn't covered and is a very important component for the security of the scheme. I also slightly elaborated on the way that the session keys are re-randomized at each hop so the sender doesn't need to blow up the packet size by including a session key for each hop. As noted in the TODO section, what's still missing here is a precise description of how the blinding factors are used to deterministically randomize for session key each hop. 

While reading I spotted one minor error (only really mentioned at the start): there isn't another "pad" key used to generate the fillers, instead just the rho key is used for this purpose. 

The other (more fundamental) error I found is the way the generation and verification of the HMACs is described:
  * An HMAC doesn't need to be computed for the final hop, it's actually all zeros. 
  * The HMAC is always computed over the "latest" _encrypted_ packet, not any of its more plaintext versions. 
  * Each hop only ever verifies the HMAC once, the only other operation related to the HMAC they perform is taking the emplaced "inner" HMAC, and using it as the "outer" HMAC for the next hop. 

I've attempted to fix the HMAC processing descriptions in line where applicable. However the diagrams that describe the HAMC computation will need to be _redone_ to reflect the fact that they're computed over the encrypted version of the packets (encrypt-then-mac).
